dd-trace-core/src/main/java/datadog/trace/core/baggage/BaggagePropagator.java

```diff
@@ -0,0 +1,192 @@
+package datadog.trace.core.baggage;
+
+import datadog.context.Context;
+import datadog.context.propagation.CarrierSetter;
+import datadog.context.propagation.CarrierVisitor;
+import datadog.context.propagation.Propagator;
+import datadog.trace.api.Config;
+import datadog.trace.bootstrap.instrumentation.api.BaggageContext;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import javax.annotation.ParametersAreNonnullByDefault;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ParametersAreNonnullByDefault
+public class BaggagePropagator implements Propagator {
+  private static final Logger log = LoggerFactory.getLogger(BaggagePropagator.class);
+  static final String BAGGAGE_KEY = "baggage";
+  private final boolean injectBaggage;
+  private final boolean extractBaggage;
+  private static final Set<Character> UNSAFE_CHARACTERS_KEY =
+      new HashSet<>(
+          Arrays.asList(
+              '\"', ',', ';', '\\', '(', ')', '/', ':', '<', '=', '>', '?', '@', '[', ']', '{',
+              '}'));
+  private static final Set<Character> UNSAFE_CHARACTERS_VALUE =
+      new HashSet<>(Arrays.asList('\"', ',', ';', '\\'));
+
+  public BaggagePropagator(boolean injectBaggage, boolean extractBaggage) {
+    this.injectBaggage = injectBaggage;
+    this.extractBaggage = extractBaggage;
+  }
+
+  private int encodeKey(String key, StringBuilder builder) {
+    return encode(key, builder, UNSAFE_CHARACTERS_KEY);
+  }
+
+  private int encodeValue(String key, StringBuilder builder) {
+    return encode(key, builder, UNSAFE_CHARACTERS_VALUE);
+  }
+
+  private int encode(String input, StringBuilder builder, Set<Character> UNSAFE_CHARACTERS) {
+    int size = 0;
+    for (int i = 0; i < input.length(); i++) {
+      char c = input.charAt(i);
+      if (UNSAFE_CHARACTERS.contains(c) || c > 126 || c <= 32) { // encode character
+        byte[] bytes = Character.toString(c).getBytes(StandardCharsets.UTF_8);
+        for (byte b : bytes) {
+          builder.append('%');
+          builder.append(encodeChar((b >> 4) & 0xF));
+          builder.append(encodeChar(b & 0xF));
+          size++;
+        }
+      } else {
+        builder.append(c);
+        size++;
+      }
+    }
+    return size;
+  }
+
+  private char encodeChar(int ascii) {
```

Comment:
This should fit in `char`
```suggestion
  private char encodeChar(char ascii) {
```

Reply:
I feel like this would not provide the desired effect because we want to cast the decimal values into their hex representation. For instance, converting the decimal representation of `0000` would result in the char representation for `null`, when we want it to be `'0'`. Let me know if this makes sense!

Reply:
Nevermind, I meant `byte` as the parameter is always (value & 0xF) so the result would fit in smaller type but int is good 👍